src/raw.rs

```diff
@@ -229,13 +229,38 @@ impl<'jvm> EnvPtr<'jvm> {
         })
     }
 
-    /// Invoke a JNI method dispatched through a virtual table lookup. Used by codegen to make most JNI calls.
+    /// Invoke a JNI method dispatched through a virtual table lookup. Used by codegen to make most JNI calls and so
+    /// must be public.
+    ///
+    /// First invokes [`FromJniValue::from_jni_value()`] before checking for a JVM exception. This allows [`Local`] drop
+    /// code to run and safely decrement the ref count before exiting early on an exception.
     ///
     /// # Safety
     ///
     /// The caller must ensure that the [`jni_sys::JNIEnv`] raw pointer is only used for this invocation.
     #[doc(hidden)]
-    pub unsafe fn invoke<F, T>(
+    pub unsafe fn invoke_checked<F, T: FromJniValue<'jvm>>(
+        self,
+        fn_field: impl FnOnce(&jni_sys::JNINativeInterface_) -> Option<F>,
+        call: impl FnOnce(*mut jni_sys::JNIEnv, F) -> T::JniValue,
+    ) -> crate::Result<'jvm, T> {
+        let value = self.invoke_unchecked(fn_field, call);
+
+        // Even if there was an exception thrown, we still need to free any non-null local ref
+        let value = T::from_jni_value(self, value);
+        self.check_exception()?;
+
+        Ok(value)
+    }
+
+    /// Invoke a JNI method dispatched through a virtual table lookup. Does *not* check for an exception and should
+    /// only be used when other mechanisms can prove the absence of an exception (e.g. a non-null return value or a
+    /// separate call to [`Self::check_exception()`]).
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that the [`jni_sys::JNIEnv`] raw pointer is only used for this invocation.
+    pub(crate) unsafe fn invoke_unchecked<F, T>(
```

Comment:
Should this be private?

Reply:
We use it in a couple of places throughout the crate (array fns, method/field lookups, string manipulation). Those places don't always return something that has a `FromJniValue` return value (e.g. method IDs or status codes) and the return value also determines whether an exception has occurred or not.

Reply:
At least it's not `pub` anymore :D